### PR TITLE
Update predicate question references when publishing

### PR DESCRIPTION
**Pull request by carolinedanzi**
### Description
1. Updates leaf nodes in the `PredicateDefinition` to use updated question IDs
2. Please scrutinize my recursive predicate tree update - it's been awhile since I've worked with tree structures!
3. Adds unit tests for `updateQuestionVersions`

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
#1001

### CODE REVIEW

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 1]
```diff
@@ -218,16 +226,56 @@ public boolean isDraft(Program program) {
...
+          Optional.of(
+              PredicateDefinition.create(
+                  updatePredicateNode(oldPredicate.rootNode()), oldPredicate.action())));
+    }
     return updatedBlock.build();
   }
 
+  // Update the referenced question IDs in all leaf nodes. Since nodes are immutable, we
+  // recursively recreate the tree with updated leaf nodes.
+  @VisibleForTesting
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:256]: Praise: ohh, I like this annotation. Thanks for adding it.

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 2]
```diff
@@ -218,16 +226,56 @@ public boolean isDraft(Program program) {
...
+  protected PredicateExpressionNode updatePredicateNode(PredicateExpressionNode current) {
```
Comment (ndmckinley) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:257]: haha you weren't kidding about the tree algorithm.  every project eventually reaches
```
case AND:
...
case OR:
...
case OP:
...
```

I'm glad we finally hit that milestone

**Review by ndmckinley — APPROVED** [review]

File: universal-application-tool-0.0.1/test/repository/VersionRepositoryTest.java [hunk 3]
```diff
@@ -73,4 +88,138 @@ public void testTransactionality() {
...
+
+    // New versions of questions
+    Question newOne = resourceCreator.insertQuestion("one");
+    newOne.addVersion(draft);
+    newOne.save();
+    Question newTwo = resourceCreator.insertQuestion("two");
+    newTwo.addVersion(draft);
+    newTwo.save();
+
+    // Build a predicate tree that covers all node types:
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/repository/VersionRepositoryTest.java:113]: Praise: thanks for adding this comment, it is super helpful.

**Review by hmhsu — APPROVED** [review]
Looks good to me.

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java (hunk 1 above)
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/repository/VersionRepository.java:256]: Thanks! It comes in handy when there are complicated methods you'd like to test!

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java (hunk 2 above)
Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/repository/VersionRepository.java:257]: I'm honored to have been involved in this historic milestone

**Pull request merged**